auth/store.go

```diff
@@ -195,6 +200,76 @@ func (as *authStore) RoleAdd(r *pb.AuthRoleAddRequest) (*pb.AuthRoleAddResponse,
 	return &pb.AuthRoleAddResponse{}, nil
 }
 
+func (as *authStore) RoleGrant(r *pb.AuthRoleGrantRequest) (*pb.AuthRoleGrantResponse, error) {
+	tx := as.be.BatchTx()
+	tx.Lock()
+	defer tx.Unlock()
+
+	_, vs := tx.UnsafeRange(authRolesBucketName, []byte(r.Name), nil, 0)
+	if len(vs) != 1 {
+		return nil, ErrRoleNotFound
+	}
+
+	role := &authpb.Role{}
+	err := role.Unmarshal(vs[0])
+	if err != nil {
+		plog.Errorf("failed to unmarshal a role %s: %s", r.Name, err)
+		return nil, err
+	}
+
+	var newPerm authpb.Permission_Type
+	if r.ReadPerm && r.WritePerm {
+		newPerm = authpb.READWRITE
+	} else if r.ReadPerm {
+		newPerm = authpb.READ
+	} else if r.WritePerm {
+		newPerm = authpb.WRITE
+	}
+
+	// TODO(mitake): clarify permission update rules
+	if !r.Prefix {
+		if !updateExistingPermission(role.KeyPermission, r.Key, newPerm) {
+			newPerm := &authpb.Permission{
+				Key:  []byte(r.Key),
+				Perm: newPerm,
+			}
+
+			role.KeyPermission = append(role.KeyPermission, newPerm)
+		}
+	} else {
+		if !updateExistingPermission(role.PrefixPermission, r.Key, newPerm) {
+			newPerm := &authpb.Permission{
+				Key:  []byte(r.Key),
+				Perm: newPerm,
+			}
+
+			role.PrefixPermission = append(role.PrefixPermission, newPerm)
+		}
+	}
+
+	marshaledRole, merr := role.Marshal()
+	if merr != nil {
+		plog.Errorf("failed to marshal updated role %s: %s", r.Name, merr)
+	}
+
+	tx.UnsafePut(authRolesBucketName, []byte(r.Name), marshaledRole)
+
+	plog.Noticef("permission of %s (read: %t, write: %t, prefix: %t) is granted to role %s", r.Key, r.ReadPerm, r.WritePerm, r.Prefix, r.Name)
```

Comment:
probably just log the permission of role X is updated to Y. Granting might be misleading (say there might not be any new permission granted).